src/pudl/analysis/allocate_net_gen.py

```diff
@@ -1598,6 +1596,80 @@ def adjust_energy_source_codes(
     return gens
 
 
+def add_missing_energy_source_codes_to_gens(gens_at_freq, gf, bf):
```

Comment:
this was the core of greg's update. all in all a great simplify-er. bc generators can have many esc & sometimes the data tables have more and different esc's, this function identifies missing esc's and adds them into the gens table.